### PR TITLE
Feature/enhanced columns

### PR DESCRIPTION
**Pull request by binarystargames**
Added a "ratio" feature to columns shortcode. Code sample/execution:

```
{{< flexcolumns ratio="3:7" >}}
**Sunrod**<br/>
*Bomb, At-Will, Standard*<br/>
*Spend 1 Reagent*
<--->
<br/>
Make an attack vs PD on 1d3+1 nearby targets.

*Hit:* The target takes radiant damage equal to your level and is Dazed (11+ save ends).

*Miss:* The target is Dazed until the start of your next turn.

*Special:* You may partially activate this as a free action (including outside of combat) to create light twice as bright as a torch for about an hour. While still lit, you may throw it to use the Bomb ability as above (it flares up and then stops working on contact).

{{< /flexcolumns >}}
```

This code becomes: https://binarystar.games/docs/36th-way/classes/list-of-classes/alchemist/#1st-level

---

I've also got a CSS rule locally to flatten columns to vertical when hugo-book gets rid of the side TOC/menu because the default hugo-book ones look not great at low width. I left it out for now but it's an option.

### CODE REVIEW

**Comment by michaeltlombardi** [comment]
Looking over this, I wonder if this isn't worth contributing back upstream to `hugo-book` directly as it's explicitly an enhancement of their shortcode. If we file an issue with them to enhance the `columns` shortcode with the `ratio` parameter, I think that will be strictly better for everyone.

If they reject the feature request / implementation, I think we can instead incorporate this as `{{ flex-columns }}` or something; I think the changeset for this is (without refactoring) too large to feasibly keep patched if the upstream implementation changes. That being said, this file does not seem to update often; 3 patches since implementation, only one since 2020.

**Comment by binarystargames** [comment]
Makes sense. I'll see if I have the energy to look into their requested process sometime this week.

**Comment by michaeltlombardi** [comment]
If you want, I'm happy to go ahead and file the issue and link to your prototype here - I really do think this is a great improvement over the upstream shortcode and hugely useful for folks!

**Comment by binarystargames** [comment]
That would be very appreciated!

**Comment by michaeltlombardi** [comment]
@binarystargames I think the upstream maintainer is unlikely to merge your contribution to the upstream. Instead of having to maintain our own implementation on top of theirs, can this be reworked to be a new shortcode called `flexcolumns`? And then we can document it and just have it be an extension here instead.

**Comment by binarystargames** [comment]
Yep, I can do that.

**Comment by binarystargames** [comment]
Added a commit that does just that, updated the overview example. I implemented it locally on my site, so that example link is current - that's using the new `{{< flexcolumns ratio="3:7" >}}` situation.

**Comment by binarystargames** [comment]
Secondary question: while I'm working on columns, how do you feel about adding a parameter to flatten them whenever the sidebar disappears? We can certainly leave it to CSS to handle too, which is what I'm currently doing for my site. Not sure how you feel about one implementation vs. the other.

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [inline layouts/shortcodes/flexcolumns.html]: nit: please add a trailing newline

**Comment by michaeltlombardi** [comment]
> Secondary question: while I'm working on columns, how do you feel about adding a parameter to flatten them whenever the sidebar disappears? We can certainly leave it to CSS to handle too, which is what I'm currently doing for my site. Not sure how you feel about one implementation vs. the other.

I think having a shortcode parameter like "flattenInMobileView" would be very useful - if people want more control over the presentation, they can ignore that parameter flag and implement their own CSS, but I don't think that's the _majority_ of our target audience.

File: layouts/shortcodes/flexcolumns.html
Comment (binarystargames) [reply layouts/shortcodes/flexcolumns.html]: Done (and I added it to the other shortcodes).

**Comment by binarystargames** [comment]
Done. 

To accomplish this, I also added `platen.scss` injected via `inject/head.html` (because I suspect to do at least a few of the things on that issues tracker we'll want somewhere to put this kind of stuff). Let me know if you'd rather I separate that out to another PR or something.

By default it uses default columns behavior, but including `flattenInMobileView="true"` adds the flattening. (I added it to Sunrod and Tanglefoot Bag from the initial demo link and nothing else so you can see the difference: https://binarystar.games/docs/36th-way/classes/list-of-classes/alchemist/#1st-level )

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [inline layouts/shortcodes/flexcolumns.html]: Something I find it useful to do for shortcodes is get/declare the variables at the top of the file to help clarify what the shortcode can expect to receive from a user / allow us to just reference the variables further down (unless they have to be pulled from the current scope, like `$ratioSplit` does). That also lets us handle when the shortcodes are called without named parameters.

You can such an example here:

https://github.com/platenio/hugo-platen/blob/64d2e468e33b53a3c14130bfe3561ba92bea5f36/layouts/shortcodes/itchio.html#L1-L11

We can _also_ use the [`default` function](https://gohugo.io/functions/default/) to set any defaults we need during parameter declaration. We could possibly use this mechanism to define a default ratio to keep from forking the code later (or not, that may not make sense here).

File: layouts/partials/platen/docs/inject/head.html
Comment (michaeltlombardi) [inline layouts/partials/platen/docs/inject/head.html]: Nit: missing trailing newline.
```suggestion
<link rel="stylesheet" href="{{ $platenStyles.RelPermalink }}" {{ template "integrity" $platenStyles }}>

```

File: assets/platen.scss
Comment (michaeltlombardi) [inline assets/platen.scss]: Nit: missing trailing newline.
```suggestion
}

```

File: layouts/partials/platen/docs/inject/head.html
Comment (michaeltlombardi) [inline layouts/partials/platen/docs/inject/head.html]: We'll need to think about this a little bit - we don't need to _solve_ it here, but if we're going to inject our own stuff into the head, we need to provide theme users with a way to add their own injections without having to shadow ours and copy-paste our code in.

**Review by michaeltlombardi — CHANGES_REQUESTED** [review]
Overall, this looks great! One minor note unrelated to the _code_, for this project if we can use rebase to update the working branch instead of merging upstream/main into it, that would be preferable - it helps keep the git history a little neater to follow.

It might also be possible to rebase and combine the first four commits into one - I think this work is really comprised of two chunks: Adding the new shortcode and extending the head injection.

Otherwise the implementation looks great and I'm stoked about this!

File: layouts/partials/platen/docs/inject/head.html
Comment (binarystargames) [reply layouts/partials/platen/docs/inject/head.html]: That's true. I'm going to aim to refactor this to inject a new file via `baseof` because that one's already non-standard from hugo-book.

File: layouts/shortcodes/flexcolumns.html
Comment (binarystargames) [reply layouts/shortcodes/flexcolumns.html]: I had no idea about .IsNamedParams or scratch! I've refactored to do this as well as use both of these, those are really good options.

**Comment by binarystargames** [comment]
Made a few changes as prescribed:

- Moved the inject out of head to baseof because we're already altering that from hugo-book.
- Refactored flexcolumns,html for improved readability/to use better tools as demonstrated

Definitely happy to rebase rather than merge in directly whenever this looks good.

File: layouts/partials/platen/platen-head.html
Comment (michaeltlombardi) [inline layouts/partials/platen/platen-head.html:2]: Commenting here to remind myself:

I have come across this pattern a couple of times now, and I think it's probably worth defining a partial for ourselves so we can reference an SCSS style link injection via something like

```go
{{ partial "platen/GetScssLink" "platen.scss" }}
```

I've been using some `utils` partials over in toroidal - it turns out partials can be used to return values instead of writing text into a file, so that might be useful.

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [inline layouts/shortcodes/flexcolumns.html]: A couple of (untested) suggestions:

```suggestion
{{-/*
  This shortcode can be used to turn markdown content into columns with a
  flexible ratio instead of being evenly distributed. It is a more powerful
  alternative to the `columns` shortcode.

  It has two optional parameters, `ratio` and `flattenInMobileView`. They can
  be specified by name or position. For example, these two shortcodes have the
  same functionality:

    {{< flexcolumns ratio="3:7" flattenInMobileView="false" >}}

    {{< flexcolumns "3:7" "false" >}}

  Note that to specify `flattenInMobileView` but not `ratio` without using
  named parameters, you **must** specify the ratio as an empty string:

    {{< flexcolumns "" "false" >}}
  
  The `ratio` parameter should be specified as a string of integers separated
  by colons (`:`), such as `1:2:1` or `1:1:3:1`. If `ratio` is not specified
  (or is set to an empty string, `""`), the columns are evenly distributed.
  For more information on flexible columns, see the docs on MDN:
  https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow
  
  If `flattenInMobileView` is specified as "true", the `flatten-in-mobile`
  class is added to the container for the columns. This causes the columns to
  collapse into a vertical stack when in mobile view to help make them more
  readable on smaller screens. If `flattenInMobileView` is not specified,
  the columns will still display horizontally, even on smaller screens.
*/-}}
{{- $ratio  := "" -}}
{{- $mobile := "" -}}

{{- if .IsNamedParams -}}
  {{- $ratio = (.Get "ratio") -}}
  {{- if eq (.Get "flattenInMobileView") "true" -}}
    {{- $mobile = " flatten-in-mobile" -}}
  {{- end -}}
{{- else -}}
  {{- $ratio = (.Get 0) -}}
  {{- if eq (.Get 1) "true" -}}
    {{- $mobile = " flatten-in-mobile" -}}
  {{- end -}}
{{- end -}}

{{- with $ratio -}}
  {{- $ratio = split . ":" -}}
{{- end -}}

<div class="book-columns flex flex-wrap{{ $mobile }}">
{{- if $ratio -}}
  {{- range $i, $val := split .Inner "<--->" -}}
    {{- if gt (len $ratio) $i}}
      <div style="flex-grow:{{ index $ratio $i }};" class="flex-even markdown-inner">
        {{- $val | $.Page.RenderString -}}
      </div>
    {{- end -}}
  {{- end -}}
{{- else -}}
  {{- range split .Inner "<--->" -}}
  <div class="flex-even markdown-inner">
    {{- . | $.Page.RenderString -}}
  </div>
  {{- end -}}
{{- end -}}
</div>
```

1. I don't _think_ I've done this so far, but I've started adding doc comments to the top of partials and shortcodes. My eventual goal is to add a little parser tool we can use to generate docs for our shortcodes/partials/etc from the first comment in a templated file. I'm also not _sure_ how the template will react with the code examples in the comment; it *should* be okay, but it might catch fire and needs manual testing.
2. While I did raise the point about `scratch` (I had been using global scratch, which is what you used here, but there's also the possibility of creating a local scratch with the [`newscratch` function][newscratch]), I think in this case there's actually some benefit to using local variables and setting them in the parameter-handling blocks:
   - It declares the parameters with their default values (in this case, empty strings) before anything else.
   - It allows us to immediately process the `flattenInMobileView` parameter to the correct value (`flatten-in-mobile` if set to `true`, empty string otherwise).
   - We can reference these multiply-used variables in the actual HTML generation without having to re-retrieve the value from the scratch map several times.
   - The use of the variables is more immediately clear to someone inspecting the template.
3. I slightly modified the class injection so that `$mobile` defaults to an empty string and is set to ` flatten-in-mobile` is set to keep from appending a trailing space to the class definition when the value is not used; this is not a big deal, but I was already poking at the code 😅 
4. I don't think the ratio splitting with the `with` directive is necessarily better than using `if` in this case, but I wanted to include it in case you weren't familiar with how `with` works; it's basically shorthand for "if this value is set to a nonzero value for this type, make it the context variable in the enclosed block" - so the inside becomes `$ratio = split . ":"` instead of `$ratio = split $ratio ":"`. Because `with` clobbers context, we can't use it in the div declaration without storing the page context in a variable; I also think changing it there would make reading the block _more_ confusing than helpful.



[newscratch]: https://gohugo.io/functions/scratch/#the-local-newscratch

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [inline layouts/shortcodes/flexcolumns.html]: Can you help me understand this block?

If I understand correctly, the ratio is split into its pieces beforehand (i.e. the ratio `1:2:3` becomes the slice `[1, 2, 3]`) and `$i` is the current index of the markdown blocks split by `<--->`. So this conditional checks to see if `$i` is greater than the length of the slice of ratio values and, if so - only if so? - adds a `div` containing the rendered content with the `flex-grow` CSS attribute set to the content's index in the slice of ratio values.

What happens when the declared ratio is `1:2:3` and there's only two sections in the markdown? When there's four sections? My brain seems to think the first item in the split markdown (index `0`?) will always be less than the length of the slice of ratio values if they're non-zero - but your code is working, which means I'm definitely misunderstanding. 😭

**Review by michaeltlombardi — COMMENTED** [review]
Some thoughts, suggestions, and questions. Also worth considering: if there are error conditions, can we handle those explicitly with the [`errorf`][errorf] function to give our users (who may not be technical at all) more helpful information than hugo's default when a template breaks?

[errorf]: https://gohugo.io/functions/errorf/

File: layouts/shortcodes/flexcolumns.html
Comment (binarystargames) [reply layouts/shortcodes/flexcolumns.html]: > So this conditional checks to see if $i is greater than the length of the slice of ratio values

gt works the other way: the slice length is greater than $i.

(Golang template scannability will be the death of me.)

File: layouts/shortcodes/flexcolumns.html
Comment (binarystargames) [reply layouts/shortcodes/flexcolumns.html]: Tested this, and aside from one change (your big block comment didn't have a space on either side, so it threw an error! Load-bearing comments are my favorite kind of error) it seems to work. As such I've put it in because it's way clearer.

File: layouts/partials/platen/platen-head.html
Comment (binarystargames) [reply layouts/partials/platen/platen-head.html:2]: Yeah, we used a similar value return with a partial with `platen/getCssClass` for class injection in `baseof`.

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [reply layouts/shortcodes/flexcolumns.html]: Ahhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

I _get_ why the functions are the way they are but I would kill for it to have logical operators implemented so it could be `if (len $ratio) > $i` 😭

**Review by michaeltlombardi — APPROVED** [review]
Okay, this looks great! I think the only dangling (non-blocking) questions are:

- is if there's any sensible error handling we can use here - maybe to check the ratio string specified, make sure that it's numbers separated by colons?
- can the `flattenInMobileView` parameter be a direct boolean instead of a string?

**Comment by binarystargames** [comment]
> sensible error handling

It's hard to believe, but this is actually pretty robust in that regard. The reason for this is how poorly formed CSS works: `flex-grow:(not a number/empty)` just doesn't get processed and falls back to `flex-grow:1`. 

At that same Alchemist link, you can see examples of:

- Needleburst: ratio="3::7"
- Immunity Shot: ratio="3:"
- Vitalizing Draught: ratio=":7
- Leadbody: ratio="3:a"
- Sliptongue: ratio="a:3"
- Incendiary Paste: ratio="3"
- Snake Venom: ratio="a"

The last two broke without a check for ":" so I've included a colon check. In all of the other cases, it just fell back to 1.

> can the `flattenInMobileView` parameter be a direct boolean instead of a string?

Sure can! The following now work:

- `{{< flexcolumns ratio="3:7" flattenInMobileView=true >}}`
- `{{< flexcolumns "3:7" true >}}`

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [inline layouts/shortcodes/flexcolumns.html]:
```suggestion
      {{< flexcolumns ratio="3:7" flattenInMobileView=false >}}
      {{< flexcolumns "3:7" false >}}
    Note that to specify `flattenInMobileView` but not `ratio` without using
    named parameters, you **must** specify the ratio as an empty string:
      {{< flexcolumns "" false >}}
```

File: layouts/shortcodes/flexcolumns.html
Comment (michaeltlombardi) [inline layouts/shortcodes/flexcolumns.html]:
```suggestion
    If `flattenInMobileView` is specified as `true`, the `flatten-in-mobile`
```

**Review by michaeltlombardi — APPROVED** [review]
Okay! That all makes sense and this looks *great*! Two minor fixes to the docs and my last (🤞🏻) question: With the way `flex-grow` works, I'm wondering if we should _warn_ when people submit an invalid ratio - probably just a single warning message like "Hey this ratio is invalid, we're going to fall back on an even ratio" and _maybe_ include an option to treat the warning as an error.

I couldn't quickly find the docs on turning warnings into errors, but I _did_ find how folks can [suppress an error](https://gohugo.io/functions/errorf/#suppress-errors). Not blocking for this PR, I'll merge tonight or tomorrow either way.

**Review by michaeltlombardi — APPROVED** [review]

**Pull request merged**
